design/design.md

```diff
@@ -0,0 +1,414 @@
+
+## Introduction
+
+This is the first design document for our Rust implementation of
+Approximate Lower Bound Arguments (ALBA) intended to guide the development
+of the library.
+We aim to create a high quality reference implementation of ALBA sufficient
+for accurately prototyping larger protocols utilizing this primitive.
+Applications that could benefit, at least in the Cardano universe, include
+Mithril, Peras, Leios and cross-chain bridges.
+These use cases will generally guide the development of our library, though
+our aim is to also allow the open-source community to study and experiment
+with ALBA.
+
+### Overview of ALBA
+ALBA is a cryptographic primitive that allows one to prove knowledge of
+some number of data elements of some kind.
+For example, one can prove knowledge of $n$ cryptographic signatures signing
+the same message, effectively constructing a multisignature scheme.
+ALBA, however, requires that there is a gap between what the prover knows and
+what the verifier is convinced the prover knows.
+Specifically, given parameters $n_f$ and $n_p$ such that $n_f < n_p$,
+the prover who possesses $n_p$ element of interest is able to convince
+the verifier that he knows $> n_f$ elements of interest.
+Hence, the lower bound argument is *approximate*.
+The larger the ratio $n_p / n_f$ is, the smaller the ALBA certificate is.
+
+ALBA can also work in the weighted regime where data elements can have
+an integer weight.
+Then the prover who possesses elements of total weight $\ge n_p$ is able to
+convince the verifier that he knows elements of total weight $> n_f$.
+Such a gap between $n_p$ and $n_f$ exists naturally in secure distributed
+systems including proof-of-stake blockchains where the amount of honest stake
+significantly exceeds the amount of malicious stake.
+
+Additionally, ALBA supports a decentralized setting where knowledge of the data
+elements is spread out across multiple provers that need to communicate to
+create a proof.
+More concretely, the provers send a message to an aggregator who in turn
+will produce a proof for the verifier.
+The primitive in this setting is called decentralized ALBA.
+Its main construction, called decentralized Telescope,
+provides a tradeoff between the proof size and the number of provers that
+communicate, leaving it up to protocol designers to make an optimal choice
+for their application.
+
+Compared to SNARKs that can solve the same problem, ALBA constructions offer
+much faster proving times: in the order of a second.
+Additionally, the decentralized Telescope ALBA construction gives rise to
+the only known multisignature scheme with a configurable tradeoff between
+proof size and (sublinear) communication complexity.
+The only other known way to achieve sublinear communication complexity is
+through a lottery as used, for example, by Algorand's sortition or the Mithril
+protocol (todo: add link).
+The lottery has slightly better communication complexity but a significantly
```

Comment:
I wonder if it makes sense to add some comparisons here - to show the overhead on proving times.